File: src/sonos/controls.ts
import { SonosClient } from './client.js';

interface PlayerState {
  volume: number;
  mute: boolean;
  playbackState: string;
  playMode: {
    repeat: string;
    shuffle: boolean;
  };
}

async function getState(client: SonosClient, room: string): Promise<PlayerState> {
  const encodedRoom = encodeURIComponent(room);
  const state = await client.request<PlayerState>(`/${encodedRoom}/state`);
  return state;
}

export async function volumeUp(client: SonosClient, room: string): Promise<number> {
  const state = await getState(client, room);
  const current = state.volume;
  // If on a multiple of 5, add 5. Otherwise round up to next multiple of 5.
  const newVolume = current % 5 === 0
    ? Math.min(100, current + 5)
    : Math.min(100, Math.ceil(current / 5) * 5);
  const encodedRoom = encodeURIComponent(room);
  await client.requestNoResponse(`/${encodedRoom}/volume/${newVolume}`);
  return newVolume;
}

export async function volumeDown(client: SonosClient, room: string): Promise<number> {
  const state = await getState(client, room);
  const current = state.volume;
  // If on a multiple of 5, subtract 5. Otherwise round down to previous multiple of 5.
  const newVolume = current % 5 === 0
    ? Math.max(0, current - 5)
    : Math.max(0, Math.floor(current / 5) * 5);
  const encodedRoom = encodeURIComponent(room);
  await client.requestNoResponse(`/${encodedRoom}/volume/${newVolume}`);
  return newVolume;
}

export async function volumeSet(client: SonosClient, room: string, level: number): Promise<number> {
  const newVolume = Math.max(0, Math.min(100, level));
  const encodedRoom = encodeURIComponent(room);
  await client.requestNoResponse(`/${encodedRoom}/volume/${newVolume}`);
  return newVolume;
}

export async function volumeHigh(client: SonosClient, room: string): Promise<number> {
  return volumeSet(client, room, 60);
}

export async function volumeLow(client: SonosClient, room: string): Promise<number> {
  return volumeSet(client, room, 10);
}

export async function getVolume(client: SonosClient, room: string): Promise<number> {
  const state = await getState(client, room);
  return state.volume;
}

export async function play(client: SonosClient, room: string): Promise<void> {
  const encodedRoom = encodeURIComponent(room);
  await client.requestNoResponse(`/${encodedRoom}/play`);
}

export async function pause(client: SonosClient, room: string): Promise<void> {
  const encodedRoom = encodeURIComponent(room);
  await client.requestNoResponse(`/${encodedRoom}/pause`);
}

export async function skip(client: SonosClient, room: string): Promise<void> {
  const encodedRoom = encodeURIComponent(room);
  await client.requestNoResponse(`/${encodedRoom}/next`);
}

export async function previous(client: SonosClient, room: string): Promise<void> {
  const encodedRoom = encodeURIComponent(room);
  await client.requestNoResponse(`/${encodedRoom}/previous`);
}

export async function restart(client: SonosClient, room: string): Promise<void> {
  const encodedRoom = encodeURIComponent(room);
  await client.requestNoResponse(`/${encodedRoom}/seek/0`);
}

export async function repeat(client: SonosClient, room: string, mode?: 'all' | 'one' | 'none'): Promise<string> {
  const encodedRoom = encodeURIComponent(room);
  if (mode) {
    await client.requestNoResponse(`/${encodedRoom}/repeat/${mode}`);
    return mode;
  }
  // Toggle through modes: none -> all -> one -> none
  const state = await getState(client, room);
  const current = state.playMode.repeat;
  const next = current === 'none' ? 'all' : current === 'all' ? 'one' : 'none';
  await client.requestNoResponse(`/${encodedRoom}/repeat/${next}`);
  return next;
}

export async function shuffle(client: SonosClient, room: string, enabled?: boolean): Promise<boolean> {
  const encodedRoom = encodeURIComponent(room);
  if (enabled !== undefined) {
    await client.requestNoResponse(`/${encodedRoom}/shuffle/${enabled ? 'on' : 'off'}`);
    return enabled;
  }
  // Toggle
  const state = await getState(client, room);
  const newState = !state.playMode.shuffle;
  await client.requestNoResponse(`/${encodedRoom}/shuffle/${newState ? 'on' : 'off'}`);
  return newState;
}

File: internal/sonos/controls.go
package sonos

import (
	"context"
	"fmt"
	"net/url"
	"strings"
)

type playerState struct {
	Volume   int `json:"volume"`
	PlayMode struct {
		Repeat  string `json:"repeat"`
		Shuffle bool   `json:"shuffle"`
	} `json:"playMode"`
}

func getState(ctx context.Context, client *Client, room string) (playerState, error) {
	var state playerState
	err := client.RequestJSON(ctx, "/"+url.PathEscape(room)+"/state", &state)
	return state, err
}

func VolumeUp(ctx context.Context, client *Client, room string) (int, error) {
	state, err := getState(ctx, client, room)
	if err != nil {
		return 0, err
	}
	current := state.Volume
	newVolume := 0
	if current%5 == 0 {
		newVolume = min(100, current+5)
	} else {
		newVolume = min(100, ((current/5)+1)*5)
	}
	err = client.RequestNoResponse(ctx, fmt.Sprintf("/%s/volume/%d", url.PathEscape(room), newVolume))
	return newVolume, err
}

func VolumeDown(ctx context.Context, client *Client, room string) (int, error) {
	state, err := getState(ctx, client, room)
	if err != nil {
		return 0, err
	}
	current := state.Volume
	newVolume := 0
	if current%5 == 0 {
		newVolume = max(0, current-5)
	} else {
		newVolume = max(0, (current/5)*5)
	}
	err = client.RequestNoResponse(ctx, fmt.Sprintf("/%s/volume/%d", url.PathEscape(room), newVolume))
	return newVolume, err
}

func VolumeSet(ctx context.Context, client *Client, room string, level int) (int, error) {
	newVolume := max(0, min(100, level))
	err := client.RequestNoResponse(ctx, fmt.Sprintf("/%s/volume/%d", url.PathEscape(room), newVolume))
	return newVolume, err
}

func VolumeHigh(ctx context.Context, client *Client, room string) (int, error) {
	return VolumeSet(ctx, client, room, 60)
}

func VolumeLow(ctx context.Context, client *Client, room string) (int, error) {
	return VolumeSet(ctx, client, room, 10)
}

func GetVolume(ctx context.Context, client *Client, room string) (int, error) {
	state, err := getState(ctx, client, room)
	if err != nil {
		return 0, err
	}
	return state.Volume, nil
}

func Play(ctx context.Context, client *Client, room string) error {
	return client.RequestNoResponse(ctx, fmt.Sprintf("/%s/play", url.PathEscape(room)))
}

func Pause(ctx context.Context, client *Client, room string) error {
	return client.RequestNoResponse(ctx, fmt.Sprintf("/%s/pause", url.PathEscape(room)))
}

func Skip(ctx context.Context, client *Client, room string) error {
	return client.RequestNoResponse(ctx, fmt.Sprintf("/%s/next", url.PathEscape(room)))
}

func Previous(ctx context.Context, client *Client, room string) error {
	return client.RequestNoResponse(ctx, fmt.Sprintf("/%s/previous", url.PathEscape(room)))
}

func Restart(ctx context.Context, client *Client, room string) error {
	return client.RequestNoResponse(ctx, fmt.Sprintf("/%s/seek/0", url.PathEscape(room)))
}

func Repeat(ctx context.Context, client *Client, room string, mode string) (string, error) {
	encodedRoom := url.PathEscape(room)
	if mode != "" {
		err := client.RequestNoResponse(ctx, fmt.Sprintf("/%s/repeat/%s", encodedRoom, mode))
		return mode, err
	}
	state, err := getState(ctx, client, room)
	if err != nil {
		return "", err
	}
	current := strings.ToLower(state.PlayMode.Repeat)
	next := "none"
	switch current {
	case "none":
		next = "all"
	case "all":
		next = "one"
	default:
		next = "none"
	}
	err = client.RequestNoResponse(ctx, fmt.Sprintf("/%s/repeat/%s", encodedRoom, next))
	return next, err
}

func Shuffle(ctx context.Context, client *Client, room string, enabled *bool) (bool, error) {
	encodedRoom := url.PathEscape(room)
	if enabled != nil {
		val := "off"
		if *enabled {
			val = "on"
		}
		err := client.RequestNoResponse(ctx, fmt.Sprintf("/%s/shuffle/%s", encodedRoom, val))
		return *enabled, err
	}
	state, err := getState(ctx, client, room)
	if err != nil {
		return false, err
	}
	newState := !state.PlayMode.Shuffle
	val := "off"
	if newState {
		val = "on"
	}
	err = client.RequestNoResponse(ctx, fmt.Sprintf("/%s/shuffle/%s", encodedRoom, val))
	return newState, err
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
